chainHead: Allow methods to be called from within a single connection context and limit connections

lexnv opened the pull request:
This PR ensures that the chainHead RPC class can be called only from within the same connection context.

The chainHead methods are now registered as raw methods. 
- https://github.com/paritytech/jsonrpsee/pull/1297
The concept of raw methods is introduced in jsonrpsee, which is an async method that exposes the connection ID:
The raw method doesn't have the concept of a blocking method. Previously blocking methods are now spawning a blocking task to handle their blocking (ie DB) access. We spawn the same number of tasks as before, however we do that explicitly.

Another approach would be implementing a RPC middleware that captures and decodes the method parameters:
- https://github.com/paritytech/polkadot-sdk/pull/3343
However, that approach is prone to errors since the methods are hardcoded by name. Performace is affected by the double deserialization that needs to happen to extract the subscription ID we'd like to limit. Once from the middleware, and once from the methods itself.

This PR paves the way to implement the chainHead connection limiter:
- https://github.com/paritytech/polkadot-sdk/issues/1505
Registering tokens (subscription ID / operation ID) on the `RpcConnections` could be extended to return an error when the maximum number of operations is reached.

While at it, have added an integration-test to ensure that chainHead methods can be called from within the same connection context.

Before this is merged, a new JsonRPC release should be made to expose the `raw-methods`:
- [x] Use jsonrpsee from crates io (blocked by: https://github.com/paritytech/jsonrpsee/pull/1297)

Closes: https://github.com/paritytech/polkadot-sdk/issues/3207


cc @paritytech/subxt-team

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5348068

niklasad1 commented on substrate/client/rpc-spec-v2/src/chain_head/chain_head.rs:
```diff
@@ -510,3 +569,26 @@ fn method_started_response(
+		let _ = tx.send(result);
+	};
+
+	executor.spawn_blocking("substrate-rpc-subscription", Some("rpc"), blocking_fut.boxed());
```
yeah right, `spawn_blocking` doesn't return a handle for the future that's why you need the extra oneshot channel

niklasad1 commented on substrate/client/rpc-spec-v2/src/common/connections.rs:
```diff
@@ -0,0 +1,60 @@
+/// Limit the RPC functionality to a single connection.
+#[derive(Default, Clone)]
+pub struct RpcConnections {
+	data: Arc<Mutex<HashMap<ConnectionId, HashSet<String>>>>,
```
could also be `Arc<Mutex<HashSet<(ConnectionId, SubscriptionId)>>>`  I guess but perhaps you have plan for it to more generic

jsdw replied:
mmm since `ConnectionId` is just a small u64, I'd just combine to `Arc<Mutex<HashSet<(ConnectionId, SubscriptionId)>>>` and that hsould be more efficient and probably small or no real extra space cost :)

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5547011

jsdw commented on substrate/client/rpc-spec-v2/src/chain_head/chain_head.rs:
```diff
@@ -209,83 +215,110 @@ where
+			.rpc_connections
+			.contains_token(connection_details.id(), &follow_subscription)
+		{
+			return ResponsePayload::success(MethodResponse::LimitReached);
```
IS this the right error to return? If this `if` is true then it means a connection is trying to call `chainHead_body` but hasn't called `chainHead_follo` yet I think?

lexnv replied:
Yep that makes sense! 

When this check is true, the current connection hasn't obtained the `follow_subscription` ID by making a call to the `chainHead_follow` first. This could mean that the user is guessing the follow subscription of another user, or that the follow subscription was leaked from another valid subscription.


> (I also can't offhand see where we do anything to limit the number of chainHead_follow subscriptions, which I assumed might exist for us to emit this error, but maybe I'm either missing something or it's not present yet :))

I'll add it in this PR since it shouldn't be that complicated, initially wanted to do a followup for this :D

jsdw replied:
One thing I didn't realise when I saw this first is that the spec actually says to return `limitReached` when the followSubscription is invalid/stale, so this all maekssense to me now (a nit here might help make clear why we return an error that feels unrelaetd at a glance eg "the spec says to return LimitReached if followSubscription is invalid or stale" :))

niklasad1 commented on substrate/client/rpc-spec-v2/Cargo.toml:
```diff
@@ -44,6 +44,7 @@ futures-util = { version = "0.3.30", default-features = false }
 rand = "0.8.5"
 
 [dev-dependencies]
+jsonrpsee = { version = "0.22", features = ["client-core", "macros", "client", "server"] }
```
"client", will add lots of dependencies AFAIU "ws-client" and "server" should be sufficient

```suggestion
jsonrpsee = { version = "0.22", features = ["ws-client", "server"] }
```

niklasad1 commented on substrate/client/rpc-spec-v2/src/chain_head/chain_head.rs:
```diff
@@ -110,6 +110,8 @@ pub struct ChainHead<BE: Backend<Block>, Block: BlockT, Client> {
 	/// The maximum number of items reported by the `chainHead_storage` before
 	/// pagination is required.
 	operation_max_storage_items: usize,
+	/// Limit the RPC functionality to a single connection.
```
I would re-phrase this, I think you are trying to say that each connection has its own state and not that the RPC functionality is limited to a single connection? 

This connection state is then used to determine whether a certain call is permitted on a particular connection such as "unstable_storage calls" or something like that...

lexnv replied:
I've rephrased the documentation a bit, let me know if it makes more sense now 🙏

niklasad1 commented on substrate/client/rpc-spec-v2/src/chain_head/chain_head.rs:
```diff
@@ -296,19 +329,33 @@ where
+			.rpc_connections
+			.contains_token(connection_details.id(), &follow_subscription)
+		{
+			return ResponsePayload::success(MethodResponse::LimitReached);
```
> This return value indicates the call couldn't be started because the server is overloaded, or that the followSubscription is invalid or stale.

I think it's quite strange to use `LimitReached` here but according to the spec...

niklasad1 commented on substrate/client/rpc-spec-v2/src/chain_head/chain_head.rs:
```diff
@@ -357,25 +404,26 @@ where
-			// Events should only by generated
-			// if the response was successfully propagated.
-			if rp_is_success.await.is_err() {
+			// This leaves on a seperate task, because the `rp` needs to be acknowledged by the
```
Wait for the server to send out the response and if it produces an error no event should be generated

lexnv replied:
Yep I do like that phrasing better, thanks!

niklasad1 commented on substrate/client/rpc-spec-v2/src/chain_head/chain_head.rs:
```diff
@@ -408,44 +463,53 @@ where
 		let fut = async move {
-			// Events should only by generated
-			// if the response was successfully propagated.
+			let event = client
```
move this below `rp_fut.await`? 

Makes not sense to me to do this storage call if  `rp_fut.await` produces an error...

lexnv replied:
Yep, that makes sense! Thanks for catching this! 🙏

niklasad1 commented on substrate/client/rpc-spec-v2/src/chain_head/tests.rs:
```diff
@@ -3297,3 +3330,120 @@ async fn storage_closest_merkle_value() {
+		.unwrap();
+
+	let mut sub: RpcClientSubscription<FollowEvent<String>> = client
+		.subscribe("chainHead_unstable_follow", rpc_params![true], "chainHead_unstable_unfollow")
```
pretty sure that you could use the type API that is generated by the jsonrpsee proc macro here when you are using a "real ws client" (not sure about name though)

```rust
client.chainhead_unstable_follow(true)
```

lexnv replied:
I completely forgot about those! I've added them since they improve a lot the code readability, thanks!

lexnv commented:
I've ended up with a slightly more generic approach to handle connections.

With this revision, we are able to:
- Allow methods to be called from as single context (chainHead methods cannot be called with a leaked subscription ID from a different connection)
- Limit the number of active `chainHead_follow` subscription (code can easily be reutilzied for transaction etc

The internal structure that handles the connections looks like:

```rust
data: Arc<Mutex<HashMap<ConnectionId, ConnectionData>>>

// Where
ConnectionData { num_tokens: usize, tokens: HashSet<String> }
```

This is because we'd need to know how many `connectionId`s are registered at any time.
The `num_tokens` counts:
- (1) tokens that are added to `tokens` (active tokens already registered)
- (2) tokens that will soon be added after the PendingSubscription is accepted

More on (2):
- We'd need to return a jsonrpsee error object when reaching the limits.
Extracted from [spec](https://github.com/paritytech/json-rpc-interface-spec/blob/main/src/api/chainHead_unstable_follow.md#possible-errors):
  > A JSON-RPC error with error code -32800 can be generated if the JSON-RPC client has already opened 2 or more chainHead_unstable_follow subscriptions.
- A `PendingSubscriptionSink` does not expose the subscription ID before being accepted (and nor should it). Therefore, this is used to "reserve in advance" some capacity. 
 - I've added an easier to use interface to abstract these concepts
 ```rust
   let Some(mut reserved: ReservedConnection) = RpcConnections.reserve_token(conenction_id) else { error };
   
   // After accepting the sink, we gain access to the subscription ID. 
   sink.accept();
   
   // Keep track of the subscription ID.
   reserved.register(follow_subscription);
   
   // RAII unregisters / unreserves on drop.
   ```

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable-int
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5572190

paritytech-cicd-pr commented:
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: cargo-clippy
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot-sdk/-/jobs/5572155

niklasad1 commented on substrate/client/rpc-spec-v2/src/chain_head/chain_head.rs:
```diff
@@ -180,12 +192,20 @@ where
+			// Ensure the current connection ID has enough space to accept a new subscription.
+			let connection_id = pending.connection_id();
+
+			let Some(reserved_token) = rpc_connections.reserve_token(connection_id) else {
```
ok, you must return a JSON-RPC error if the limit is exceeded and that's why you look up the state in the hashmap before "subscription accept is called"?

Too bad to lock the mutex twice for that...  but no way around that I guess

niklasad1 replied:
Can we merge `subscriptions` and `connections`? The subscription ID is stored in both and would be nice to merge them is possible

niklasad1 commented on substrate/client/rpc-spec-v2/src/chain_head/chain_head.rs:
```diff
@@ -198,6 +218,8 @@ where
 			};
 			debug!(target: LOG_TARGET, "[follow][id={:?}] Subscription accepted", sub_id);
 
+			let _registered_token = reserved_token.register(sub_id.clone());
```
doesn't return anything?!

```suggestion
			reserved_token.register(sub_id.clone());
```

niklasad1 replied:
RAII, when it dropped it's removed from connections

niklasad1 commented on substrate/client/rpc-spec-v2/src/common/connections.rs:
```diff
@@ -0,0 +1,228 @@
+	/// Register a token for the given connection.
+	///
+	/// Users should call [`Self::reserve_token`] before calling this method.
+	fn register_token(&self, connection_id: ConnectionId, token: String) {
```
this should return RegisteredToken?

niklasad1 commented on substrate/client/rpc-spec-v2/src/common/connections.rs:
```diff
@@ -0,0 +1,228 @@
+	sync::Arc,
+};
+
+/// Limit the RPC functionality to a single connection.
```
```suggestion
/// Connection state which keeps track whether a connection exist and
/// the number of concurrent operations.
```

niklasad1 commented on substrate/client/rpc-spec-v2/src/common/connections.rs:
```diff
@@ -0,0 +1,228 @@
+	///
+	/// # Note
+	///
+	/// Because a pending subscription sink does not expose the future subscription ID,
```
I wonder if we can make it explicit that token == subscription ID?

It may be hard to understand this comment about the subscription sink otherwise

lexnv replied:
Indeed the `token` name is a bit misleading. This is the subscription ID in this case, however it will be the transaction `OperatationId` for the tx API 🤔 

Maybe `identifier` sounds better? 🤔

niklasad1 commented on substrate/client/rpc-spec-v2/src/common/connections.rs:
```diff
@@ -0,0 +1,228 @@
+			return;
+		}
+
+		entry.tokens.insert(token);
```
we assume that subscription ID is unique so should be unlikely for duplicate subscription IDs here

niklasad1 commented on substrate/client/rpc-spec-v2/src/common/connections.rs:
```diff
@@ -0,0 +1,228 @@
+}
+
+#[cfg(test)]
+mod tests {
```
Would useful with another test "is_empty" that connection IDs are cleaned from the HashMaps etc.

The code looks go to me otherwise

lexnv replied:
Added a few extra checks to ensure we clean-up everything now, thanks!

jsdw commented on substrate/client/rpc-spec-v2/src/chain_head/subscription/mod.rs:
```diff
@@ -43,30 +64,55 @@ impl<Block: BlockT, BE: Backend<Block>> SubscriptionManagement<Block, BE> {
+	/// # Note
+	///
+	/// Used for testing.
+	#[doc(hidden)]
```
If only used for testing, you could have `#[cfg(test)]` here instead to guarantee that (and make it super obvious) :)

jsdw commented on substrate/client/rpc-spec-v2/src/chain_head/tests.rs:
```diff
@@ -3297,3 +3350,176 @@ async fn storage_closest_merkle_value() {
+		_ => panic!("Unexpected subscription ID"),
+	};
+
+	// Cannot make a call from a different connection context.
+	let _response = ChainHeadApiClient::<String>::chain_head_unstable_unpin(
+		&second_client,
+		first_sub_id.clone(),
+		crate::chain_head::api::ListOrValue::Value(finalized_hash.clone()),
+	)
+	.await
+	.unwrap();
```
This code produces no error I guess (we just unwrap it), so what do we learn from it?

jsdw replied:
Ah I see; we are unpinning the block from the wrong subscription and that hsould have no effect!

jsdw commented on substrate/client/rpc-spec-v2/src/chain_head/tests.rs:
```diff
@@ -3297,3 +3350,176 @@ async fn storage_closest_merkle_value() {
+		_ => panic!("Unexpected subscription ID"),
+	};
+
+	// Cannot make a call from a different connection context.
```
Maybe tweak the comment to "trying to unpin from a different conneciton will have no effect"?

jsdw commented on substrate/client/rpc-spec-v2/src/common/connections.rs:
```diff
@@ -0,0 +1,258 @@
+	/// For transactionBroadcast, this represents the operation ID.
+	/// For transaction, this is empty and the number of active calls is tracked by
+	/// [`Self::num_identifiers`].
+	identifiers: HashSet<String>,
```
Am I right to assume that we are not worries about subscription IDs and operation IDs colliding at all in our implementation here?

lexnv replied:
Yep, IIRC, we'll either use this for chainHead with subscriptions ID only; or for the tx with operation ID only

jsdw commented on substrate/client/rpc-spec-v2/src/common/connections.rs:
```diff
@@ -0,0 +1,258 @@
+		let mut data = self.data.lock();
+
+		let entry = data.entry(connection_id).or_insert_with(ConnectionData::default);
+		entry.num_identifiers = entry.num_identifiers.saturating_sub(1);
```
Do we ever clear out entries completely when the number goes to 0?

jsdw commented on substrate/client/rpc-spec-v2/src/common/connections.rs:
```diff
@@ -0,0 +1,258 @@
+			connection_data.num_identifiers = connection_data.num_identifiers.saturating_sub(1);
+
+			if connection_data.num_identifiers == 0 {
+				data.remove(&connection_id);
```
Ah, I guess this is whewre we clean up :)

jsdw reviewed (approved):
Looks good to me; only very small comments in the end :)
